Sources/Regex/ASTPrinting.swift

```diff
@@ -0,0 +1,65 @@
+
+// MARK: - Printing
+
+/// AST entities can be pretty-printed or dumped
+///
+/// Alternative: just use `description` for pretty-print
+/// and `debugDescription` for dump
+public protocol _ASTPrintable:
```

Comment:
If this module is only visible to the compiler (and not to the end developer), any reason we need an underscore for this?

Reply:
It's available to a number of clients:

- The compiler
- The stdlib
- SourceKit (in some TBD form)
- SwiftSyntax (in some TBD form)
- Future: LSP
- Future: 3rd party tools and libraries

Basically, we're writing compilers, which are better designed as libraries. The AST is the frontend's API (or more specifically AST and a corresponding ASTAction protocol of some sort).

It's also underscored because I know for sure this isn't a great design long-term but I like it a little better than `Custom[Debug]StringConvertible` for now. It's easier to change and adapt over time. It's a utility/tooling thing, but we lack access-controlled protocol conformances. This certainly isn't API in the sense that AST is, but might be in the future for pretty-printing and dumping, interface TBD.

I'm good going either way, but I want to start refining our use of `public` in the `Regex` module.

Reply:
I haven't thought this through too deeply, so take this with a grain of salt. We're looking at converting this into a functional prototype really soon, integrated with the compiler and everything. I want to start getting clearer on what's a useful shortcut for prototyping and what's explicitly being prototyped or demonstrated. I'm definitely not prototyping AST printing, but I _am_ prototyping/demonstrating our AST. I wonder if a leading underscore convention for this distinction would be helpful.